src/storage/index/IndexExecutor.inl

```diff
@@ -115,16 +117,96 @@ cpp2::ErrorCode IndexExecutor<RESP>::checkReturnColumns(const std::vector<std::s
     return cpp2::ErrorCode::SUCCEEDED;
 }
 
+// TODO (sky) : String range scan was disabled graph layer. it is not support in storage layer.
+template <typename RESP>
+std::pair<std::string, std::string>
+IndexExecutor<RESP>::makeScanPair(PartitionID partId, IndexID indexId) {
+    std::string beginStr = NebulaKeyUtils::indexPrefix(partId, indexId);
+    std::string endStr = NebulaKeyUtils::indexPrefix(partId, indexId);
+    const auto& fields = index_->get_fields();
+    for (const auto& field : fields) {
+        auto item = scanItems_.find(field.get_name());
+        if (item == scanItems_.end()) {
+            break;
+        }
+        // here need check the value type, maybe different data types appear.
+        // for example:
+        // index (c1 double)
+        // where c1 > abs(1) , FunctionCallExpression->eval(abs(1))
+        // should be cast type from int to double.
+        bool suc = true;
+        if (item->second.beginBound_.rel_ != RelationType::kNull) {
+            suc = NebulaKeyUtils::checkAndCastVariant(field.get_type().type,
+                                                      item->second.beginBound_.val_);
+        }
+        if (item->second.endBound_.rel_ != RelationType::kNull) {
+            suc = NebulaKeyUtils::checkAndCastVariant(field.get_type().type,
+                                                      item->second.endBound_.val_);
+        }
+        if (!suc) {
+            VLOG(1) << "Unknown VariantType";
+            return {};
+        }
+        auto pair = normalizeScanPair(field, (*item).second);
+        beginStr.append(pair.first);
+        endStr.append(pair.second);
+    }
+    return std::make_pair(beginStr, endStr);
+}
+
+template <typename RESP>
+std::pair<std::string, std::string>
+IndexExecutor<RESP>::normalizeScanPair(const nebula::cpp2::ColumnDef& field,
+                                       const ScanBound& item) {
+    std::string begin, end;
+    auto type = field.get_type().type;
+    // if begin == end, means the scan is equivalent scan.
+    if (item.beginBound_.rel_ != RelationType::kGTRel &&
+        item.endBound_.rel_ != RelationType::kGTRel &&
+        item.beginBound_.rel_ != RelationType::kNull &&
+        item.endBound_.rel_ != RelationType::kNull &&
+        item.beginBound_.val_ == item.endBound_.val_) {
+        begin = end = NebulaKeyUtils::encodeVariant(item.beginBound_.val_);
```

Comment:
em....    
if begin.val == end.val
It seems  only the begin.rel == end.rel is a valid operation, right?

Reply:
Yes, that's right.